### ISSUE TITLE
ngtools/webpack resolve fails for packages with a "browser" field

### ISSUE THREAD

**Issue by jonrimmer** [angular/angular-cli#7540]
### Bug Report or Feature Request (mark with an `x`)
```
- [x] bug report
- [ ] feature request
```

### Versions.
1.6.2

### Repro steps.
See the following simple reproduction:

https://github.com/jonrimmer/ngtools-webpack-is-broken

```
npm i
npm run webpack
```

### The log given by the failure.

```
WARNING in ./node_modules/ws/lib/BufferUtil.js
Module not found: Error: Can't resolve 'bufferutil' in '/Users/jon/Dev/ngtools-webpack-is-broken/node_modules/ws/lib'
 @ ./node_modules/ws/lib/BufferUtil.js 12:15-36
 @ ./node_modules/ws/lib/Sender.js
 @ ./node_modules/ws/index.js
 @ ./node_modules/atmosphere.js/index.js
 @ ./src/main.ts
 @ multi ./src/main.ts atmosphere.js @angular/core @angular/common

WARNING in ./node_modules/ws/lib/Validation.js
Module not found: Error: Can't resolve 'utf-8-validate' in '/Users/jon/Dev/ngtools-webpack-is-broken/node_modules/ws/lib'
 @ ./node_modules/ws/lib/Validation.js 12:16-41
 @ ./node_modules/ws/lib/Receiver.js
 @ ./node_modules/ws/index.js
 @ ./node_modules/atmosphere.js/index.js
 @ ./src/main.ts
 @ multi ./src/main.ts atmosphere.js @angular/core @angular/common

ERROR in ./node_modules/xmlhttprequest/lib/XMLHttpRequest.js
Module not found: Error: Can't resolve 'child_process' in '/Users/jon/Dev/ngtools-webpack-is-broken/node_modules/xmlhttprequest/lib'
 @ ./node_modules/xmlhttprequest/lib/XMLHttpRequest.js 15:12-36
 @ ./node_modules/atmosphere.js/index.js
 @ ./src/main.ts
 @ multi ./src/main.ts atmosphere.js @angular/core @angular/common

ERROR in ./node_modules/options/lib/options.js
Module not found: Error: Can't resolve 'fs' in '/Users/jon/Dev/ngtools-webpack-is-broken/node_modules/options/lib'
 @ ./node_modules/options/lib/options.js 6:9-22
 @ ./node_modules/ws/lib/WebSocketServer.js
 @ ./node_modules/ws/index.js
 @ ./node_modules/atmosphere.js/index.js
 @ ./src/main.ts
 @ multi ./src/main.ts atmosphere.js @angular/core @angular/common

ERROR in ./node_modules/xmlhttprequest/lib/XMLHttpRequest.js
Module not found: Error: Can't resolve 'fs' in '/Users/jon/Dev/ngtools-webpack-is-broken/node_modules/xmlhttprequest/lib'
 @ ./node_modules/xmlhttprequest/lib/XMLHttpRequest.js 16:9-22
 @ ./node_modules/atmosphere.js/index.js
 @ ./src/main.ts
 @ multi ./src/main.ts atmosphere.js @angular/core @angular/common

ERROR in ./node_modules/ws/lib/WebSocketServer.js
Module not found: Error: Can't resolve 'tls' in '/Users/jon/Dev/ngtools-webpack-is-broken/node_modules/ws/lib'
 @ ./node_modules/ws/lib/WebSocketServer.js 15:10-24
 @ ./node_modules/ws/index.js
 @ ./node_modules/atmosphere.js/index.js
 @ ./src/main.ts
 @ multi ./src/main.ts atmosphere.js @angular/core @angular/common

```

### Desired functionality.

The example imports the npm package `atmosphere.js`, which has two entry points specified in its `package.json`: `"main": "index.js"` for node.js, and `"browser": "./lib/browser.js"` for the browser. Using `ngtools/webpack` causes the wrong one to be used. This then causes the build to fail, because the node.js entry point tries to `require()` various node packages.

This does not happen when using JIT compilation.

**Comment by jonrimmer** [comment]
@hansl This appears to have been caused by d1d2aa0. The paths plugin is not properly restricting where it resolves things from. E.g. if I have a path mapping in my `tsconfig.json` as follows:

```
  "paths": {
    "*": ["./src/*]
  }
```

And I do the following in my app:

```
import * as atmosphere from 'atmosphere.js';
```

The paths plugin is using TypeScript to resolve the module `atmosphere.js` using TypeScript's resolution strategy. However, rather than restricting its search to within `./src/`, as the mapping says, it is also looking in `node_modules`. It finds a match here, but since TypeScript's resolution strategy does not support [mainFields](https://webpack.js.org/configuration/resolve/#resolve-mainfields), it resolves to the main entry point, instead of the browser entry point. This in turn causes that file to be loaded and its node-only `require(...)`s to be resolved.

**Comment by jskrzypek** [comment]
This bug has been exacerbated now by https://github.com/angular/angular-cli/commit/eb4c9e72ed09c6b544f2dddd038099cfb6d914c8 as this which causes webpack's normal strategies for reading the `browser` field to not be applied to any `.js` dependencies that match the paths patterns as well. Specifically we had issues with fabric.js and socket.io-client. (See my colleague's issue on the matter: #8878).

To solve this we were forced to be much more selective about the patterns we use in the tsconfig `paths` for @ancular/cli projects, and we were forced to use separate tsconfigs for our cli apps and our libs compiled with `ngc`, as neither works completely on it's own.

**Comment by jskrzypek** [comment]
@filipesilva @hansl, I was curious about the source of this error and manually changed the installed version of `@ngtools/webpack/src/paths-plugin.js` in my project according to @jonrimmer's [comment](https://github.com/angular/angular-cli/commit/d1d2aa0ccf646bf377916272b5a28b251bd4c41f#r23978061) like so:

```diff
                 }
-                const moduleResolver = ts.resolveModuleName(request.request, request.contextInfo.issuer, this._compilerOptions, this._host);
+                const moduleResolver = ts.resolveModuleName(newRequestStr, request.contextInfo.issuer, this._compilerOptions, this._host);
                 let moduleFilePath = moduleResolver.resolvedModule
                     && moduleResolver.resolvedModule.resolvedFileName;
```

This fixed the issue I was seeing without the need to modify my tsconfig's `paths` to be more selective, which may have unintended side effects.

**Comment by clydin** [comment]
This should be completely corrected in the upcoming 7.0 which uses webpack's configured resolution strategy for path mapping.
